awq/models/mixtral.py

```diff
@@ -0,0 +1,139 @@
+import tqdm
+from typing import List, Tuple
+from .base import BaseAWQForCausalLM
+from awq.utils.fused_utils import fuse_qkv
+from awq.modules.fused.block import MixtralBlock
+from awq.modules.fused.model import LlamaLikeModel
+from transformers.models.mixtral.modeling_mixtral import (
+    MixtralDecoderLayer as OldmixtralDecoderLayer,
+    MixtralForCausalLM as OldmixtralForCausalLM
+)
+from awq.modules.fused.mlp import QuantFusedMLP
+from awq.modules.fused.norm import FasterTransformerRMSNorm
+
+class MixtralAWQForCausalLM(BaseAWQForCausalLM):
+    layer_type = "MixtralDecoderLayer"
+    max_new_tokens_key = "max_position_embeddings"
+
+    @staticmethod
+    def get_model_layers(model: OldmixtralForCausalLM):
+        return model.model.layers
+    
+    @staticmethod
+    def get_act_for_scaling(module):
+        return dict(
+            is_scalable=False
+        )
+    
+    @staticmethod
+    def move_embed(model: OldmixtralForCausalLM, device: str):
+        model.model.embed_tokens = model.model.embed_tokens.to(device)
+    
+    @staticmethod
+    def get_layers_for_scaling(module: OldmixtralDecoderLayer, input_feat, module_kwargs):
+        layers = []
+
+        # attention input
+        layers.append(dict(
+            prev_op=module.input_layernorm,
+            layers=[module.self_attn.q_proj,
+                    module.self_attn.k_proj, module.self_attn.v_proj],
+            inp=input_feat['self_attn.q_proj'],
+            module2inspect=module.self_attn, kwargs=module_kwargs,
+        ))
+
+        # attention out
+        if module.self_attn.v_proj.weight.shape == module.self_attn.o_proj.weight.shape:
+            layers.append(dict(
+                prev_op=module.self_attn.v_proj,
+                layers=[module.self_attn.o_proj],
+                inp=input_feat['self_attn.o_proj'],
+            ))
+        
+        # first expert
+        layers.append(dict(
+            prev_op=module.post_attention_layernorm,
+            layers=[module.block_sparse_moe.experts[0].w1,
+                    module.block_sparse_moe.experts[0].w3],
+            inp=input_feat[f'block_sparse_moe.experts.{0}.w1'],
+            module2inspect=module.block_sparse_moe.experts[0],
+        ))
+
+        # remaining experts
+        experts = module.block_sparse_moe.experts
+        for i, expert in enumerate(experts[1:]):
+            expert_index = i+1
+
+            layers.append(dict(
+                prev_op=expert.w3,
+                layers=[expert.w1, expert.w3],
```

Comment:
why `prev_op` and `layers` all contain `expert.w3`

Reply:
it means don't apply scaling for the remaining experts

Reply:
Feel free to test out other ways of quantizing Mixtral and make a contribution. I am in the process of figuring out how it can be scaled without losing much in perplexity.